In [1]:
import os
import joblib
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import TfidfVectorizer
from lightgbm import LGBMClassifier
from sklearn.metrics import accuracy_score, classification_report
import spacy
import optuna
import mlflow
import mlflow.sklearn

# MLflow / DagShub tracking setup
os.environ["MLFLOW_TRACKING_USERNAME"] = "Roy7721"
os.environ["MLFLOW_TRACKING_PASSWORD"] = "bc912b7d58bd2aca05abdc192e010414493a3886"
mlflow.set_tracking_uri("https://dagshub.com/Roy7721/yt_comment_analysis.mlflow")
mlflow.set_experiment("Custom Features - LightGBM HP Tuning")

C:\Users\Dell\AppData\Roaming\Python\Python314\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


<Experiment: artifact_location='mlflow-artifacts:/5a6385f1f2c24471add67c02cd88d561', creation_time=1784594071756, effective_trace_archival_retention=None, experiment_id='12', last_update_time=1784594071756, lifecycle_stage='active', name='Custom Features - LightGBM HP Tuning', tags={}, trace_location=None, workspace='default'>

In [2]:
# Load dataset (same file + same cleaning as the LogReg custom-features notebook)
dataset = pd.read_csv('dataset.csv')

# Drop rows with NaN values
cleaned_dataset = dataset.dropna()

In [3]:
# Separate features and target
X_cleaned = cleaned_dataset['clean_comment']
y_cleaned = cleaned_dataset['category']

# 80-20 split - IDENTICAL to the LogReg custom-features notebook
# (same random_state=42, same test_size, no stratify) -> identical rows -> fair comparison
X_train_cleaned, X_test_cleaned, y_train_cleaned, y_test_cleaned = train_test_split(
    X_cleaned, y_cleaned, test_size=0.2, random_state=42
)

In [4]:
# Load spacy language model for POS tagging
nlp = spacy.load('en_core_web_sm')

In [ ]:
# All 17 universal POS tags, in a FIXED order.
# Looping over this (instead of set(pos_tags)) guarantees every comment produces the
# same 17 POS columns in the same order -> train, test, and future data always line up.
UNIVERSAL_POS = ['ADJ', 'ADP', 'ADV', 'AUX', 'CCONJ', 'DET', 'INTJ', 'NOUN', 'NUM',
                 'PART', 'PRON', 'PROPN', 'PUNCT', 'SCONJ', 'SYM', 'VERB', 'X']

# Function to extract custom features
def extract_custom_features(text):
    doc = nlp(text)
    word_list = [token.text for token in doc]

    # 1. Comment Length (number of characters)
    comment_length = len(text)

    # 2. Word Count
    word_count = len(word_list)

    # 3. Average Word Length
    avg_word_length = sum(len(word) for word in word_list) / word_count if word_count > 0 else 0

    # 4. Unique Word Count
    unique_word_count = len(set(word_list))

    # 5. Lexical Diversity
    lexical_diversity = unique_word_count / word_count if word_count > 0 else 0

    # 6. Count of POS Tags
    pos_count = len([token.pos_ for token in doc])

    # 7. Proportion of POS Tags - loop over the FIXED list (not set()) so every comment
    #    yields all 17 columns in the same order; tags not present come out as 0.
    pos_tags = [token.pos_ for token in doc]
    if word_count > 0:
        pos_proportion = {tag: pos_tags.count(tag) / word_count for tag in UNIVERSAL_POS}
    else:
        pos_proportion = {tag: 0 for tag in UNIVERSAL_POS}

    return {
        'comment_length': comment_length,
        'word_count': word_count,
        'avg_word_length': avg_word_length,
        'unique_word_count': unique_word_count,
        'lexical_diversity': lexical_diversity,
        'pos_count': pos_count,
        **pos_proportion  # Flattening the POS proportions
    }

In [6]:
# Apply the custom feature extraction
train_custom_features = pd.DataFrame([extract_custom_features(text) for text in X_train_cleaned])
test_custom_features = pd.DataFrame([extract_custom_features(text) for text in X_test_cleaned])

In [7]:
# Replace NaN values in POS tag proportions with 0
train_custom_features.fillna(0, inplace=True)
test_custom_features.fillna(0, inplace=True)

In [8]:
# TF-IDF - bigram, max_features=10000, FIT ONLY on train
ngram_range = (1, 2)
max_features = 10000
tfidf = TfidfVectorizer(ngram_range=ngram_range, max_features=max_features)
X_train_tfidf = tfidf.fit_transform(X_train_cleaned)  # Fit ONLY on train
X_test_tfidf = tfidf.transform(X_test_cleaned)        # Transform test

In [9]:
# Convert TF-IDF to DataFrame
X_train_tfidf_df = pd.DataFrame(X_train_tfidf.toarray(), columns=tfidf.get_feature_names_out())
X_test_tfidf_df = pd.DataFrame(X_test_tfidf.toarray(), columns=tfidf.get_feature_names_out())

In [10]:
# Combine TF-IDF and custom features
X_train_combined = pd.concat([X_train_tfidf_df.reset_index(drop=True), train_custom_features.reset_index(drop=True)], axis=1)
X_test_combined = pd.concat([X_test_tfidf_df.reset_index(drop=True), test_custom_features.reset_index(drop=True)], axis=1)

# Align test columns to train columns (order AND set).
# The spaCy POS-proportion columns come from set(pos_tags), so train/test can end up in
# different column orders -> LightGBM/sklearn feature-name mismatch at predict.
# reindex fixes the order, adds any POS column missing from test (as 0), and drops extras.
X_test_combined = X_test_combined.reindex(columns=X_train_combined.columns, fill_value=0)

In [11]:
# LightGBM is scale-invariant -> NO scaler needed (unlike the LogReg notebook)
X_train = X_train_combined
X_test = X_test_combined
y_train = y_train_cleaned
y_test = y_test_cleaned

In [12]:
# Function to log results in MLflow
def log_mlflow(model_name, model, X_train, X_test, y_train, y_test, params, trial_number, log_model=False):
    with mlflow.start_run():
        # Log model type and trial number
        mlflow.set_tag("mlflow.runName", f"Trial_{trial_number}_{model_name}_")
        mlflow.set_tag("experiment_type", "algorithm_comparison")

        # Log algorithm name as a parameter
        mlflow.log_param("algo_name", model_name)
        mlflow.log_param("vectorizer_type", "TfidfVectorizer")
        mlflow.log_param("ngram_range", str(ngram_range))
        mlflow.log_param("vectorizer_max_features", max_features)

        # Log hyperparameters
        for key, value in params.items():
            mlflow.log_param(key, value)

        # Train model
        model.fit(X_train, y_train)
        y_pred = model.predict(X_test)

        # Log accuracy
        accuracy = accuracy_score(y_test, y_pred)
        mlflow.log_metric("accuracy", accuracy)

        # Log classification report
        classification_rep = classification_report(y_test, y_pred, output_dict=True)
        for label, metrics in classification_rep.items():
            if isinstance(metrics, dict):
                for metric, value in metrics.items():
                    mlflow.log_metric(f"{label}_{metric}", value)

        # Only log the model artifact for the best run to stay within DagShub free tier limits
        if log_model:
            # Persist locally first so a long train is never lost to a logging failure
            joblib.dump(model, f"{model_name}_best_model.joblib")
            try:
                mlflow.sklearn.log_model(
                    model, f"{model_name}_model",
                    serialization_format=mlflow.sklearn.SERIALIZATION_FORMAT_CLOUDPICKLE,
                )
            except Exception as e:
                print(f"[warn] model artifact logging failed (metrics/params + joblib file are safe): {e}")

        return accuracy

In [13]:
# Step 6: Optuna objective function for LightGBM
def objective_lightgbm(trial):
    # Hyperparameter space to explore
    n_estimators = trial.suggest_int('n_estimators', 100, 1000)
    learning_rate = trial.suggest_float('learning_rate', 1e-4, 1e-1, log=True)
    max_depth = trial.suggest_int('max_depth', 3, 15)
    num_leaves = trial.suggest_int('num_leaves', 20, 150)
    min_child_samples = trial.suggest_int('min_child_samples', 10, 100)
    colsample_bytree = trial.suggest_float('colsample_bytree', 0.5, 1.0)
    subsample = trial.suggest_float('subsample', 0.5, 1.0)
    reg_alpha = trial.suggest_float('reg_alpha', 1e-4, 10.0, log=True)   # L1 regularization
    reg_lambda = trial.suggest_float('reg_lambda', 1e-4, 10.0, log=True)  # L2 regularization
    # Class imbalance handling - let Optuna decide whether balancing helps
    class_weight = trial.suggest_categorical('class_weight', [None, 'balanced'])

    # Log trial parameters
    params = {
        'n_estimators': n_estimators,
        'learning_rate': learning_rate,
        'max_depth': max_depth,
        'num_leaves': num_leaves,
        'min_child_samples': min_child_samples,
        'colsample_bytree': colsample_bytree,
        'subsample': subsample,
        'reg_alpha': reg_alpha,
        'reg_lambda': reg_lambda,
        'class_weight': class_weight
    }

    # Create LightGBM model (no scaler needed - trees are scale-invariant)
    model = LGBMClassifier(n_estimators=n_estimators,
                           learning_rate=learning_rate,
                           max_depth=max_depth,
                           num_leaves=num_leaves,
                           min_child_samples=min_child_samples,
                           colsample_bytree=colsample_bytree,
                           subsample=subsample,
                           reg_alpha=reg_alpha,
                           reg_lambda=reg_lambda,
                           class_weight=class_weight,
                           random_state=42,
                           verbose=-1)

    # Log each trial as a separate run in MLflow
    accuracy = log_mlflow("LightGBM", model, X_train, X_test, y_train, y_test, params, trial.number)

    return accuracy

In [14]:
# Step 7: Run Optuna for LightGBM, log the best model, and plot the importance of each parameter
def run_optuna_experiment():
    study = optuna.create_study(direction="maximize")
    study.optimize(objective_lightgbm, n_trials=50)  # 50 trials (matches the LogReg notebook)

    # Get the best parameters
    best_params = study.best_params
    best_model = LGBMClassifier(n_estimators=best_params['n_estimators'],
                                learning_rate=best_params['learning_rate'],
                                max_depth=best_params['max_depth'],
                                num_leaves=best_params['num_leaves'],
                                min_child_samples=best_params['min_child_samples'],
                                colsample_bytree=best_params['colsample_bytree'],
                                subsample=best_params['subsample'],
                                reg_alpha=best_params['reg_alpha'],
                                reg_lambda=best_params['reg_lambda'],
                                class_weight=best_params['class_weight'],
                                random_state=42,
                                verbose=-1)

    # Log the best model with MLflow (only this run uploads the model artifact)
    log_mlflow("LightGBM", best_model, X_train, X_test, y_train, y_test, best_params, "Best", log_model=True)

    # Plot parameter importance
    optuna.visualization.plot_param_importances(study).show()

    # Plot optimization history
    optuna.visualization.plot_optimization_history(study).show()

    # Return the best model so it can be inspected outside this function
    return best_model

In [19]:
# Run the experiment for LightGBM
best_model = run_optuna_experiment()

[I 2026-07-21 18:54:12,063] A new study created in memory with name: no-name-6c5d6013-da95-4d8f-a14c-666e16895002


🏃 View run Trial_0_LightGBM_ at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12/runs/6254a98c103b45f4a86dd0726dcf740d
🧪 View experiment at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12


[I 2026-07-21 18:58:44,875] Trial 0 finished with value: 0.760125460248193 and parameters: {'n_estimators': 404, 'learning_rate': 0.03505690255420824, 'max_depth': 4, 'num_leaves': 20, 'min_child_samples': 53, 'colsample_bytree': 0.9067624187994126, 'subsample': 0.5471132771207625, 'reg_alpha': 0.00040455816980270283, 'reg_lambda': 0.0921027850530837, 'class_weight': None}. Best is trial 0 with value: 0.760125460248193.


🏃 View run Trial_1_LightGBM_ at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12/runs/5eb2df8b70e94fec974e7e5277c1aa4b
🧪 View experiment at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12


[I 2026-07-21 19:04:33,313] Trial 1 finished with value: 0.802945588435838 and parameters: {'n_estimators': 909, 'learning_rate': 0.030343166923543434, 'max_depth': 7, 'num_leaves': 81, 'min_child_samples': 74, 'colsample_bytree': 0.7349400488583051, 'subsample': 0.5402766661292986, 'reg_alpha': 0.000120733785767025, 'reg_lambda': 0.0007948199191232308, 'class_weight': None}. Best is trial 1 with value: 0.802945588435838.


🏃 View run Trial_2_LightGBM_ at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12/runs/6477785d7d6f46848cd1b73c15ea6767
🧪 View experiment at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12


[I 2026-07-21 19:08:01,183] Trial 2 finished with value: 0.7222146461202782 and parameters: {'n_estimators': 376, 'learning_rate': 0.015263513949699926, 'max_depth': 6, 'num_leaves': 116, 'min_child_samples': 78, 'colsample_bytree': 0.7519704453056759, 'subsample': 0.5623302667511244, 'reg_alpha': 0.0006854151331675238, 'reg_lambda': 0.0016855398247455282, 'class_weight': 'balanced'}. Best is trial 1 with value: 0.802945588435838.


🏃 View run Trial_3_LightGBM_ at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12/runs/3e67704a30b74e50b13a3628443bb5cf
🧪 View experiment at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12


[I 2026-07-21 19:12:14,312] Trial 3 finished with value: 0.6312559661802809 and parameters: {'n_estimators': 528, 'learning_rate': 0.000131040652925132, 'max_depth': 13, 'num_leaves': 69, 'min_child_samples': 96, 'colsample_bytree': 0.8469867948629384, 'subsample': 0.5793662874992733, 'reg_alpha': 0.00024160155557126793, 'reg_lambda': 6.851288455658672, 'class_weight': 'balanced'}. Best is trial 1 with value: 0.802945588435838.


🏃 View run Trial_4_LightGBM_ at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12/runs/48c5a4c777194e9ba5d378b1ebfcaad7
🧪 View experiment at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12


[I 2026-07-21 19:14:39,493] Trial 4 finished with value: 0.793945179326333 and parameters: {'n_estimators': 135, 'learning_rate': 0.09256649476186425, 'max_depth': 8, 'num_leaves': 125, 'min_child_samples': 48, 'colsample_bytree': 0.7960243489472854, 'subsample': 0.5629566047179217, 'reg_alpha': 0.49731513820402673, 'reg_lambda': 0.00015357155951337248, 'class_weight': None}. Best is trial 1 with value: 0.802945588435838.
C:\Users\Dell\AppData\Roaming\Python\Python314\site-packages\sklearn\metrics\_classification.py:1879: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", result.shape[0])
C:\Users\Dell\AppData\Roaming\Python\Python314\site-packages\sklearn\metrics\_classification.py:1879: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` par

🏃 View run Trial_5_LightGBM_ at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12/runs/a55319d21a084aec9ef173659375e999
🧪 View experiment at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12


[I 2026-07-21 19:19:00,638] Trial 5 finished with value: 0.4331105959361789 and parameters: {'n_estimators': 548, 'learning_rate': 0.00011752597138081228, 'max_depth': 15, 'num_leaves': 40, 'min_child_samples': 47, 'colsample_bytree': 0.7632941964742486, 'subsample': 0.5636264730055288, 'reg_alpha': 0.002149894778887344, 'reg_lambda': 0.00010406054525594557, 'class_weight': None}. Best is trial 1 with value: 0.802945588435838.
C:\Users\Dell\AppData\Roaming\Python\Python314\site-packages\sklearn\metrics\_classification.py:1879: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", result.shape[0])
C:\Users\Dell\AppData\Roaming\Python\Python314\site-packages\sklearn\metrics\_classification.py:1879: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division

🏃 View run Trial_6_LightGBM_ at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12/runs/5c5589c6cdd547b2a0f31f705953758d
🧪 View experiment at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12


[I 2026-07-21 19:23:55,941] Trial 6 finished with value: 0.6177553525160234 and parameters: {'n_estimators': 500, 'learning_rate': 0.0004375976261806562, 'max_depth': 12, 'num_leaves': 52, 'min_child_samples': 59, 'colsample_bytree': 0.6662149350477643, 'subsample': 0.8684559362876685, 'reg_alpha': 0.0012819518331304979, 'reg_lambda': 0.06130774443103862, 'class_weight': None}. Best is trial 1 with value: 0.802945588435838.


🏃 View run Trial_7_LightGBM_ at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12/runs/1340eff41caa41a4bf5ffbe3b24f3ac7
🧪 View experiment at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12


[I 2026-07-21 19:27:27,111] Trial 7 finished with value: 0.5754807036683486 and parameters: {'n_estimators': 520, 'learning_rate': 0.0004249354447678098, 'max_depth': 4, 'num_leaves': 124, 'min_child_samples': 54, 'colsample_bytree': 0.753613700412306, 'subsample': 0.9755578017169335, 'reg_alpha': 0.0022243049437703365, 'reg_lambda': 0.00035204041878564627, 'class_weight': 'balanced'}. Best is trial 1 with value: 0.802945588435838.


🏃 View run Trial_8_LightGBM_ at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12/runs/b0c0accd943f439086f9c70a5e343bb6
🧪 View experiment at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12


[I 2026-07-21 19:31:59,021] Trial 8 finished with value: 0.7858993590617755 and parameters: {'n_estimators': 865, 'learning_rate': 0.01791845301795599, 'max_depth': 14, 'num_leaves': 129, 'min_child_samples': 91, 'colsample_bytree': 0.6110205681637875, 'subsample': 0.581425093814244, 'reg_alpha': 4.147175013284183, 'reg_lambda': 7.126360762834239, 'class_weight': None}. Best is trial 1 with value: 0.802945588435838.


🏃 View run Trial_9_LightGBM_ at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12/runs/8fd364378a2c42639790d08b92204118
🧪 View experiment at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12


[I 2026-07-21 19:39:05,669] Trial 9 finished with value: 0.7376244374744306 and parameters: {'n_estimators': 990, 'learning_rate': 0.003348345128436815, 'max_depth': 14, 'num_leaves': 76, 'min_child_samples': 89, 'colsample_bytree': 0.8077334165143808, 'subsample': 0.9175764179584405, 'reg_alpha': 0.005447296472767061, 'reg_lambda': 1.8241249204569827, 'class_weight': None}. Best is trial 1 with value: 0.802945588435838.


🏃 View run Trial_10_LightGBM_ at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12/runs/f9442ccc499149f59fd3c32a77971d4d
🧪 View experiment at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12


[I 2026-07-21 19:51:59,118] Trial 10 finished with value: 0.7027137597163507 and parameters: {'n_estimators': 762, 'learning_rate': 0.0022445562725775373, 'max_depth': 10, 'num_leaves': 93, 'min_child_samples': 12, 'colsample_bytree': 0.9893909399786092, 'subsample': 0.7397866973701257, 'reg_alpha': 0.08775696008269937, 'reg_lambda': 0.006787934920772049, 'class_weight': 'balanced'}. Best is trial 1 with value: 0.802945588435838.


🏃 View run Trial_11_LightGBM_ at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12/runs/a06d667d278040e1a069cd571128f384
🧪 View experiment at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12


[I 2026-07-21 19:54:33,234] Trial 11 finished with value: 0.7598527205782081 and parameters: {'n_estimators': 164, 'learning_rate': 0.06559789139829231, 'max_depth': 8, 'num_leaves': 100, 'min_child_samples': 32, 'colsample_bytree': 0.5027969407706115, 'subsample': 0.6768180160080594, 'reg_alpha': 8.605699612192858, 'reg_lambda': 0.0011710505598282962, 'class_weight': None}. Best is trial 1 with value: 0.802945588435838.


🏃 View run Trial_12_LightGBM_ at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12/runs/b9ca9da4f2f04bcb8ec669cc6cc15a10
🧪 View experiment at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12


[I 2026-07-21 19:56:47,013] Trial 12 finished with value: 0.791763261966453 and parameters: {'n_estimators': 148, 'learning_rate': 0.09815443350649862, 'max_depth': 8, 'num_leaves': 149, 'min_child_samples': 74, 'colsample_bytree': 0.6634863157019558, 'subsample': 0.6793052525435499, 'reg_alpha': 0.16663778003938676, 'reg_lambda': 0.00011506001344640051, 'class_weight': None}. Best is trial 1 with value: 0.802945588435838.


🏃 View run Trial_13_LightGBM_ at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12/runs/b3037604b2b546bfaca9266f9739d769
🧪 View experiment at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12


[I 2026-07-21 20:01:55,468] Trial 13 finished with value: 0.7508523114687031 and parameters: {'n_estimators': 742, 'learning_rate': 0.00872695907054071, 'max_depth': 7, 'num_leaves': 146, 'min_child_samples': 32, 'colsample_bytree': 0.8769543744172801, 'subsample': 0.5092128369258508, 'reg_alpha': 0.03230852002944679, 'reg_lambda': 0.007665766660845083, 'class_weight': None}. Best is trial 1 with value: 0.802945588435838.


🏃 View run Trial_14_LightGBM_ at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12/runs/dda3983275bd4dd7a01a8005119014ec
🧪 View experiment at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12


[I 2026-07-21 20:05:58,460] Trial 14 finished with value: 0.793126960316378 and parameters: {'n_estimators': 295, 'learning_rate': 0.040627059154665524, 'max_depth': 10, 'num_leaves': 95, 'min_child_samples': 69, 'colsample_bytree': 0.6828010162429365, 'subsample': 0.6492710276160741, 'reg_alpha': 0.6110452507740968, 'reg_lambda': 0.0009298778758120263, 'class_weight': None}. Best is trial 1 with value: 0.802945588435838.


🏃 View run Trial_15_LightGBM_ at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12/runs/c4a2277998074e4ab62392a7741f7cea
🧪 View experiment at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12


[I 2026-07-21 20:16:00,036] Trial 15 finished with value: 0.7483976544388381 and parameters: {'n_estimators': 661, 'learning_rate': 0.006440145908825714, 'max_depth': 10, 'num_leaves': 74, 'min_child_samples': 37, 'colsample_bytree': 0.9466093785127885, 'subsample': 0.6362155627948406, 'reg_alpha': 0.014117294003065361, 'reg_lambda': 0.009603887982966153, 'class_weight': None}. Best is trial 1 with value: 0.802945588435838.


🏃 View run Trial_16_LightGBM_ at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12/runs/057cf5771be64053bd0b177f5b07d116
🧪 View experiment at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12


[I 2026-07-21 20:24:24,414] Trial 16 finished with value: 0.812764216555298 and parameters: {'n_estimators': 982, 'learning_rate': 0.024293119870993318, 'max_depth': 6, 'num_leaves': 110, 'min_child_samples': 12, 'colsample_bytree': 0.5792312063895599, 'subsample': 0.7741919885834203, 'reg_alpha': 1.0899429531879383, 'reg_lambda': 0.00042603183308174145, 'class_weight': None}. Best is trial 16 with value: 0.812764216555298.


🏃 View run Trial_17_LightGBM_ at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12/runs/901294df8d6b44c2bf88711aaeb6df72
🧪 View experiment at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12


[I 2026-07-21 20:32:28,260] Trial 17 finished with value: 0.7628528569480431 and parameters: {'n_estimators': 974, 'learning_rate': 0.022144827228160014, 'max_depth': 3, 'num_leaves': 106, 'min_child_samples': 15, 'colsample_bytree': 0.5067879493838581, 'subsample': 0.7841597482567297, 'reg_alpha': 1.6087306909653267, 'reg_lambda': 0.04132216177919379, 'class_weight': None}. Best is trial 16 with value: 0.812764216555298.


🏃 View run Trial_18_LightGBM_ at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12/runs/c09a03241dd147638226e04f595ceae8
🧪 View experiment at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12


[I 2026-07-21 20:40:01,662] Trial 18 finished with value: 0.6833492431474157 and parameters: {'n_estimators': 874, 'learning_rate': 0.0023359269940364394, 'max_depth': 6, 'num_leaves': 84, 'min_child_samples': 21, 'colsample_bytree': 0.5734328857932138, 'subsample': 0.8012312971274246, 'reg_alpha': 0.00012842363056449747, 'reg_lambda': 0.0030574144774210233, 'class_weight': 'balanced'}. Best is trial 16 with value: 0.812764216555298.


🏃 View run Trial_19_LightGBM_ at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12/runs/24dc852e96a24637814b9ec40b8cf033
🧪 View experiment at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12


[I 2026-07-21 20:43:39,073] Trial 19 finished with value: 0.7320332742397382 and parameters: {'n_estimators': 866, 'learning_rate': 0.007446072957222123, 'max_depth': 5, 'num_leaves': 49, 'min_child_samples': 63, 'colsample_bytree': 0.5707129506013429, 'subsample': 0.8530769545071067, 'reg_alpha': 0.030442366214799198, 'reg_lambda': 0.0005728130756281113, 'class_weight': None}. Best is trial 16 with value: 0.812764216555298.


🏃 View run Trial_20_LightGBM_ at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12/runs/5d9ddc6d55854ba1bd71860378d69fab
🧪 View experiment at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12


[I 2026-07-21 20:47:55,711] Trial 20 finished with value: 0.8269466793945179 and parameters: {'n_estimators': 775, 'learning_rate': 0.03845043777255551, 'max_depth': 6, 'num_leaves': 111, 'min_child_samples': 24, 'colsample_bytree': 0.7040700116611793, 'subsample': 0.7252582983892446, 'reg_alpha': 0.17071782646518358, 'reg_lambda': 0.4052725267054083, 'class_weight': None}. Best is trial 20 with value: 0.8269466793945179.


🏃 View run Trial_21_LightGBM_ at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12/runs/ce98752af0cb454eb19e3c4fe7860c9d
🧪 View experiment at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12


[I 2026-07-21 20:52:36,104] Trial 21 finished with value: 0.8243556525296605 and parameters: {'n_estimators': 761, 'learning_rate': 0.041671586807026026, 'max_depth': 6, 'num_leaves': 109, 'min_child_samples': 24, 'colsample_bytree': 0.7024012421894567, 'subsample': 0.7335195526244124, 'reg_alpha': 1.3510216867953853, 'reg_lambda': 0.33465218136902825, 'class_weight': None}. Best is trial 20 with value: 0.8269466793945179.


🏃 View run Trial_22_LightGBM_ at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12/runs/19233a751be0411aae6011baf7fadba2
🧪 View experiment at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12


[I 2026-07-21 20:59:42,956] Trial 22 finished with value: 0.8228555843447429 and parameters: {'n_estimators': 731, 'learning_rate': 0.04731672449701685, 'max_depth': 5, 'num_leaves': 112, 'min_child_samples': 23, 'colsample_bytree': 0.6257412548399252, 'subsample': 0.7377876225974341, 'reg_alpha': 0.7010152383598481, 'reg_lambda': 0.3557593333016373, 'class_weight': None}. Best is trial 20 with value: 0.8269466793945179.


🏃 View run Trial_23_LightGBM_ at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12/runs/0317d3a647394e8784dfed80ba2e2e03
🧪 View experiment at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12


[I 2026-07-21 21:03:43,710] Trial 23 finished with value: 0.802400109095868 and parameters: {'n_estimators': 679, 'learning_rate': 0.05753538174850886, 'max_depth': 3, 'num_leaves': 136, 'min_child_samples': 24, 'colsample_bytree': 0.7126527488005776, 'subsample': 0.7230201622184205, 'reg_alpha': 0.182625872436045, 'reg_lambda': 0.4194242717284629, 'class_weight': None}. Best is trial 20 with value: 0.8269466793945179.


🏃 View run Trial_24_LightGBM_ at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12/runs/c16f4082e90b4287ba4711b9a155650a
🧪 View experiment at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12


[I 2026-07-21 21:07:37,634] Trial 24 finished with value: 0.7434883403791082 and parameters: {'n_estimators': 637, 'learning_rate': 0.01430994270203728, 'max_depth': 5, 'num_leaves': 109, 'min_child_samples': 23, 'colsample_bytree': 0.629765445011523, 'subsample': 0.729212079076899, 'reg_alpha': 2.3881952529366894, 'reg_lambda': 0.3968010300510976, 'class_weight': None}. Best is trial 20 with value: 0.8269466793945179.


🏃 View run Trial_25_LightGBM_ at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12/runs/8be1f48d4ef84f659f7de1bbf4e30ee2
🧪 View experiment at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12


[I 2026-07-21 21:12:03,404] Trial 25 finished with value: 0.8328105822991955 and parameters: {'n_estimators': 787, 'learning_rate': 0.060077868166864945, 'max_depth': 5, 'num_leaves': 116, 'min_child_samples': 38, 'colsample_bytree': 0.6962793490874111, 'subsample': 0.8292198588854258, 'reg_alpha': 0.3964082222149267, 'reg_lambda': 0.3864152832163617, 'class_weight': None}. Best is trial 25 with value: 0.8328105822991955.


🏃 View run Trial_26_LightGBM_ at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12/runs/35d345a8881a41528f263b4540467c55
🧪 View experiment at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12


[I 2026-07-21 21:16:43,461] Trial 26 finished with value: 0.833765171144143 and parameters: {'n_estimators': 790, 'learning_rate': 0.05953891185333248, 'max_depth': 9, 'num_leaves': 137, 'min_child_samples': 41, 'colsample_bytree': 0.6943338465876195, 'subsample': 0.8154524442491013, 'reg_alpha': 0.26497365335517503, 'reg_lambda': 1.1349968019453902, 'class_weight': 'balanced'}. Best is trial 26 with value: 0.833765171144143.


🏃 View run Trial_27_LightGBM_ at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12/runs/f866411a54a34eb390962f150627aa87
🧪 View experiment at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12


[I 2026-07-21 21:21:30,519] Trial 27 finished with value: 0.8352652393290604 and parameters: {'n_estimators': 842, 'learning_rate': 0.07031988383905727, 'max_depth': 9, 'num_leaves': 138, 'min_child_samples': 41, 'colsample_bytree': 0.8104792946572928, 'subsample': 0.8372406252622299, 'reg_alpha': 0.3003905146576322, 'reg_lambda': 1.302172224859495, 'class_weight': 'balanced'}. Best is trial 27 with value: 0.8352652393290604.


🏃 View run Trial_28_LightGBM_ at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12/runs/5545384e214846edb599052e89744c19
🧪 View experiment at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12


[I 2026-07-21 21:25:45,647] Trial 28 finished with value: 0.8401745533887904 and parameters: {'n_estimators': 832, 'learning_rate': 0.06579560301311387, 'max_depth': 11, 'num_leaves': 138, 'min_child_samples': 39, 'colsample_bytree': 0.8133181100339337, 'subsample': 0.8484243225444615, 'reg_alpha': 0.32916988759784627, 'reg_lambda': 1.6388150292982968, 'class_weight': 'balanced'}. Best is trial 28 with value: 0.8401745533887904.


🏃 View run Trial_29_LightGBM_ at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12/runs/46edf8e004794b989daf6a348c73e24a
🧪 View experiment at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12


[I 2026-07-21 21:29:18,715] Trial 29 finished with value: 0.8273557888994955 and parameters: {'n_estimators': 821, 'learning_rate': 0.09059116568755202, 'max_depth': 11, 'num_leaves': 138, 'min_child_samples': 45, 'colsample_bytree': 0.8360081836407661, 'subsample': 0.9085214768685714, 'reg_alpha': 0.05309345038227142, 'reg_lambda': 1.7889197571749351, 'class_weight': 'balanced'}. Best is trial 28 with value: 0.8401745533887904.


🏃 View run Trial_30_LightGBM_ at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12/runs/a472a7d3ccef40518b4e1426e779b00f
🧪 View experiment at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12


[I 2026-07-21 21:32:35,048] Trial 30 finished with value: 0.774580662757398 and parameters: {'n_estimators': 694, 'learning_rate': 0.011713817698237572, 'max_depth': 9, 'num_leaves': 138, 'min_child_samples': 40, 'colsample_bytree': 0.9076845722940745, 'subsample': 0.9132329045908453, 'reg_alpha': 0.25400265558817897, 'reg_lambda': 1.4734083007329632, 'class_weight': 'balanced'}. Best is trial 28 with value: 0.8401745533887904.


🏃 View run Trial_31_LightGBM_ at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12/runs/c126e2969f264df5947f8adab6767361
🧪 View experiment at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12


[I 2026-07-21 21:35:50,852] Trial 31 finished with value: 0.8394927042138279 and parameters: {'n_estimators': 817, 'learning_rate': 0.0629239318051224, 'max_depth': 11, 'num_leaves': 121, 'min_child_samples': 39, 'colsample_bytree': 0.8027051986354797, 'subsample': 0.8283567568122975, 'reg_alpha': 0.38034954452714353, 'reg_lambda': 0.1693958418512354, 'class_weight': 'balanced'}. Best is trial 28 with value: 0.8401745533887904.


🏃 View run Trial_32_LightGBM_ at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12/runs/05b013cad49b4ed5b7fc0884401f23dc
🧪 View experiment at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12


[I 2026-07-21 21:40:27,290] Trial 32 finished with value: 0.8464475657984454 and parameters: {'n_estimators': 935, 'learning_rate': 0.028423424681821317, 'max_depth': 11, 'num_leaves': 131, 'min_child_samples': 31, 'colsample_bytree': 0.7867848770634716, 'subsample': 0.8747799623338763, 'reg_alpha': 0.07628060357797721, 'reg_lambda': 0.1332106173031934, 'class_weight': 'balanced'}. Best is trial 32 with value: 0.8464475657984454.


🏃 View run Trial_33_LightGBM_ at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12/runs/d253aaded78a40e9940ac43b83e75a38
🧪 View experiment at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12


[I 2026-07-21 21:44:38,313] Trial 33 finished with value: 0.8506750306832128 and parameters: {'n_estimators': 918, 'learning_rate': 0.030407755468826486, 'max_depth': 12, 'num_leaves': 130, 'min_child_samples': 30, 'colsample_bytree': 0.7862563104213793, 'subsample': 0.8712514927712355, 'reg_alpha': 0.0974769484269205, 'reg_lambda': 0.1366578699355978, 'class_weight': 'balanced'}. Best is trial 33 with value: 0.8506750306832128.


🏃 View run Trial_34_LightGBM_ at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12/runs/2f65a58b81b540628fb43a103178a809
🧪 View experiment at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12


[I 2026-07-21 21:48:37,520] Trial 34 finished with value: 0.8456293467884903 and parameters: {'n_estimators': 926, 'learning_rate': 0.02903715769239081, 'max_depth': 12, 'num_leaves': 122, 'min_child_samples': 32, 'colsample_bytree': 0.7812877355326959, 'subsample': 0.885296957654992, 'reg_alpha': 0.07726285129821645, 'reg_lambda': 0.10461827269829203, 'class_weight': 'balanced'}. Best is trial 33 with value: 0.8506750306832128.


🏃 View run Trial_35_LightGBM_ at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12/runs/5c24ed6f71b64e0288e19f61ae8e2fbc
🧪 View experiment at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12


[I 2026-07-21 21:53:24,829] Trial 35 finished with value: 0.8468566753034229 and parameters: {'n_estimators': 956, 'learning_rate': 0.026239878388114788, 'max_depth': 12, 'num_leaves': 130, 'min_child_samples': 31, 'colsample_bytree': 0.7669967088279628, 'subsample': 0.8791322055473565, 'reg_alpha': 0.012772377718368524, 'reg_lambda': 0.13950048140408558, 'class_weight': 'balanced'}. Best is trial 33 with value: 0.8506750306832128.


🏃 View run Trial_36_LightGBM_ at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12/runs/38073ffe9cb942cb9b7034c7504b8bf3
🧪 View experiment at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12


[I 2026-07-21 21:58:51,152] Trial 36 finished with value: 0.8498568116732579 and parameters: {'n_estimators': 934, 'learning_rate': 0.02691122170245552, 'max_depth': 13, 'num_leaves': 130, 'min_child_samples': 29, 'colsample_bytree': 0.7728645569497017, 'subsample': 0.9653057076064232, 'reg_alpha': 0.009887310174739884, 'reg_lambda': 0.02150061913765767, 'class_weight': 'balanced'}. Best is trial 33 with value: 0.8506750306832128.


🏃 View run Trial_37_LightGBM_ at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12/runs/a2bb4d8b1577418a8001c52ec665ae89
🧪 View experiment at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12


[I 2026-07-21 22:07:40,213] Trial 37 finished with value: 0.7586253920632756 and parameters: {'n_estimators': 925, 'learning_rate': 0.004294250982341538, 'max_depth': 13, 'num_leaves': 130, 'min_child_samples': 29, 'colsample_bytree': 0.7359825788024172, 'subsample': 0.9782278798146851, 'reg_alpha': 0.010375969525919114, 'reg_lambda': 0.01674019005208256, 'class_weight': 'balanced'}. Best is trial 33 with value: 0.8506750306832128.


🏃 View run Trial_38_LightGBM_ at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12/runs/fefbc47960564e02bbb4668a5cec9223
🧪 View experiment at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12


[I 2026-07-21 22:17:02,017] Trial 38 finished with value: 0.8243556525296605 and parameters: {'n_estimators': 935, 'learning_rate': 0.013081820948334539, 'max_depth': 13, 'num_leaves': 147, 'min_child_samples': 16, 'colsample_bytree': 0.855530747330749, 'subsample': 0.925176893917906, 'reg_alpha': 0.008420563738727155, 'reg_lambda': 0.035065880867403997, 'class_weight': 'balanced'}. Best is trial 33 with value: 0.8506750306832128.


🏃 View run Trial_39_LightGBM_ at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12/runs/7b473a05a51b47b7bffd60dd8c36d5d7
🧪 View experiment at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12


[I 2026-07-21 22:22:03,888] Trial 39 finished with value: 0.8161734624301105 and parameters: {'n_estimators': 897, 'learning_rate': 0.019716466541130812, 'max_depth': 12, 'num_leaves': 130, 'min_child_samples': 53, 'colsample_bytree': 0.7740366935276857, 'subsample': 0.9504650825119995, 'reg_alpha': 0.004466098632687791, 'reg_lambda': 0.14241214869986143, 'class_weight': 'balanced'}. Best is trial 33 with value: 0.8506750306832128.


🏃 View run Trial_40_LightGBM_ at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12/runs/dd098da79c88458dbd9b34fd746e6593
🧪 View experiment at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12


[I 2026-07-21 22:29:48,357] Trial 40 finished with value: 0.8523114687031229 and parameters: {'n_estimators': 937, 'learning_rate': 0.028634453866698585, 'max_depth': 15, 'num_leaves': 117, 'min_child_samples': 29, 'colsample_bytree': 0.8788509902686811, 'subsample': 0.954337205538961, 'reg_alpha': 0.02502355006139102, 'reg_lambda': 0.023604311288344712, 'class_weight': 'balanced'}. Best is trial 40 with value: 0.8523114687031229.


🏃 View run Trial_41_LightGBM_ at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12/runs/a07dc42dc77a4005a3287e0eb26c9205
🧪 View experiment at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12


[I 2026-07-21 22:37:31,342] Trial 41 finished with value: 0.8519023591981454 and parameters: {'n_estimators': 949, 'learning_rate': 0.02761166582629805, 'max_depth': 14, 'num_leaves': 128, 'min_child_samples': 30, 'colsample_bytree': 0.8746548154547574, 'subsample': 0.9493995947286186, 'reg_alpha': 0.02118805693470823, 'reg_lambda': 0.020360706765862213, 'class_weight': 'balanced'}. Best is trial 40 with value: 0.8523114687031229.


🏃 View run Trial_42_LightGBM_ at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12/runs/b0a48f2380d5411a9456f238346be0ae
🧪 View experiment at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12


[I 2026-07-21 22:46:51,743] Trial 42 finished with value: 0.8204009273148779 and parameters: {'n_estimators': 995, 'learning_rate': 0.009951599164471041, 'max_depth': 15, 'num_leaves': 120, 'min_child_samples': 27, 'colsample_bytree': 0.8895162078646215, 'subsample': 0.9978791413404998, 'reg_alpha': 0.020849013360704143, 'reg_lambda': 0.02208733770149931, 'class_weight': 'balanced'}. Best is trial 40 with value: 0.8523114687031229.


🏃 View run Trial_43_LightGBM_ at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12/runs/07dc75088b2c41cfb738cd9d2e2621fd
🧪 View experiment at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12


[I 2026-07-21 22:53:55,212] Trial 43 finished with value: 0.8609027683076503 and parameters: {'n_estimators': 959, 'learning_rate': 0.0322975539986836, 'max_depth': 14, 'num_leaves': 102, 'min_child_samples': 19, 'colsample_bytree': 0.9342710606329229, 'subsample': 0.9443188557579818, 'reg_alpha': 0.003021526575654611, 'reg_lambda': 0.003532663186327057, 'class_weight': 'balanced'}. Best is trial 43 with value: 0.8609027683076503.


🏃 View run Trial_44_LightGBM_ at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12/runs/5d02de402d5143c4ad9696c7554738c2
🧪 View experiment at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12


[I 2026-07-21 23:04:09,875] Trial 44 finished with value: 0.8418109914087004 and parameters: {'n_estimators': 892, 'learning_rate': 0.018674312203469868, 'max_depth': 14, 'num_leaves': 98, 'min_child_samples': 18, 'colsample_bytree': 0.9327939377439478, 'subsample': 0.9491355646568711, 'reg_alpha': 0.0007024846174786055, 'reg_lambda': 0.004499455452652125, 'class_weight': 'balanced'}. Best is trial 43 with value: 0.8609027683076503.


🏃 View run Trial_45_LightGBM_ at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12/runs/93aa880ec66349d0ac5f4ced0eb5d50e
🧪 View experiment at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12


[I 2026-07-21 23:12:57,927] Trial 45 finished with value: 0.8607663984726579 and parameters: {'n_estimators': 905, 'learning_rate': 0.03219676051562872, 'max_depth': 15, 'num_leaves': 89, 'min_child_samples': 10, 'colsample_bytree': 0.9574152680835801, 'subsample': 0.9422232122212425, 'reg_alpha': 0.0030897920300712975, 'reg_lambda': 0.0021824164539841384, 'class_weight': 'balanced'}. Best is trial 43 with value: 0.8609027683076503.


🏃 View run Trial_46_LightGBM_ at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12/runs/c4f4e01a07414a058d1c17721799d2bd
🧪 View experiment at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12


[I 2026-07-21 23:24:11,954] Trial 46 finished with value: 0.8629483158325378 and parameters: {'n_estimators': 902, 'learning_rate': 0.03525642576883964, 'max_depth': 15, 'num_leaves': 88, 'min_child_samples': 11, 'colsample_bytree': 0.9975747444801832, 'subsample': 0.9294604717135944, 'reg_alpha': 0.001713556836930048, 'reg_lambda': 0.0022080042696159116, 'class_weight': 'balanced'}. Best is trial 46 with value: 0.8629483158325378.


🏃 View run Trial_47_LightGBM_ at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12/runs/946beb7fa94b42f5961be5315b730efe
🧪 View experiment at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12


[I 2026-07-21 23:41:15,045] Trial 47 finished with value: 0.6889404063821083 and parameters: {'n_estimators': 869, 'learning_rate': 0.0005300945707995307, 'max_depth': 15, 'num_leaves': 88, 'min_child_samples': 12, 'colsample_bytree': 0.9841541261825442, 'subsample': 0.9376514727298074, 'reg_alpha': 0.0023271627924843163, 'reg_lambda': 0.0018919738150605715, 'class_weight': 'balanced'}. Best is trial 46 with value: 0.8629483158325378.


🏃 View run Trial_48_LightGBM_ at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12/runs/182bc75858674af0b419c549d59eeb30
🧪 View experiment at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12


[I 2026-07-21 23:56:02,117] Trial 48 finished with value: 0.8446747579435429 and parameters: {'n_estimators': 995, 'learning_rate': 0.016752417952338033, 'max_depth': 14, 'num_leaves': 61, 'min_child_samples': 10, 'colsample_bytree': 0.9513705585592978, 'subsample': 0.9969073219843763, 'reg_alpha': 0.0005545032063340721, 'reg_lambda': 0.0022509121035195006, 'class_weight': 'balanced'}. Best is trial 46 with value: 0.8629483158325378.


🏃 View run Trial_49_LightGBM_ at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12/runs/0ec2238f8bf549299a748b7ad87a0300
🧪 View experiment at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12


[I 2026-07-22 00:09:59,049] Trial 49 finished with value: 0.8622664666575753 and parameters: {'n_estimators': 964, 'learning_rate': 0.035820546921949276, 'max_depth': 14, 'num_leaves': 103, 'min_child_samples': 19, 'colsample_bytree': 0.9985977535303042, 'subsample': 0.8949879097000155, 'reg_alpha': 0.0011716743462494815, 'reg_lambda': 0.005037659652254979, 'class_weight': 'balanced'}. Best is trial 46 with value: 0.8629483158325378.
2026/07/22 00:21:51 WARNING mlflow.models.model: `artifact_path` is deprecated. Please use `name` instead.
2026/07/22 00:22:04 WARNING mlflow.sklearn: Saving scikit-learn models in the pickle or cloudpickle format requires exercising caution because these formats rely on Python's object serialization mechanism, which can execute arbitrary code during deserialization. The recommended safe alternative is the 'skops' format. For more information, see: https://scikit-learn.org/stable/model_persistence.html


🏃 View run Trial_Best_LightGBM_ at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12/runs/3e702f5b0f9741ae90318e20532d4433
🧪 View experiment at: https://dagshub.com/Roy7721/yt_comment_analysis.mlflow/#/experiments/12


ImportError: Tried to import 'plotly' but failed. Please make sure that the package is installed correctly to use this feature. Actual error: No module named 'plotly'.

In [ ]:
best_model

# 📌 Best Params — LightGBM + Custom Features

**Run:** 2026-07-21 18:58 → 2026-07-22 00:10 (~5h 11m) · 50/50 Optuna trials completed
**Experiment (MLflow):** `Custom Features - LightGBM HP Tuning`
**Features:** TF-IDF bigram (1,2), `max_features=10000` + spaCy custom features (23 cols) → 10,023 total
**Split:** 80/20, `random_state=42`, no stratify (identical to the LogReg custom-features notebook)

## 🏆 Best result — Trial 46

**Test accuracy: `0.8629`**

```python
best_params_lightgbm = {
    'n_estimators':      902,
    'learning_rate':     0.03525642576883964,
    'max_depth':         15,
    'num_leaves':        88,
    'min_child_samples': 11,
    'colsample_bytree':  0.9975747444801832,
    'subsample':         0.9294604717135944,
    'reg_alpha':         0.001713556836930048,
    'reg_lambda':        0.0022080042696159116,
    'class_weight':      'balanced',
    # fixed (not tuned)
    'random_state':      42,
    'verbose':           -1,
}
```

## Top 5 trials

| Rank | Trial | Accuracy | learning_rate | max_depth | num_leaves | min_child_samples | class_weight |
|------|-------|----------|---------------|-----------|------------|-------------------|--------------|
| 1 | **46** | **0.8629** | 0.0353 | 15 | 88  | 11 | balanced |
| 2 | 49 | 0.8623 | 0.0358 | 14 | 103 | 19 | balanced |
| 3 | 43 | 0.8609 | 0.0323 | 14 | 102 | 19 | balanced |
| 4 | 45 | 0.8608 | 0.0322 | 15 | 89  | 10 | balanced |
| 5 | 40 | 0.8523 | 0.0286 | 15 | 117 | 29 | balanced |

## Observations

- **`class_weight='balanced'` won decisively** — every one of the top 5 trials used it. Trials with `class_weight=None` plateaued around 0.83.
- The winning region is consistent: **learning_rate ≈ 0.03**, **max_depth 14–15**, **num_leaves ~90–105**, **low min_child_samples (10–20)**, **high colsample_bytree/subsample (~0.93–1.0)**, **weak regularization** (`reg_alpha`/`reg_lambda` ≈ 0.002–0.005).
- Low learning rates (< 0.001) were disastrous (0.43–0.69) — the model never converged within the estimator budget.
- **Custom features helped:** untuned LightGBM + custom features scored 0.8523; tuned reaches **0.8629**.

## Notes

- The run ended with an `ImportError: No module named 'plotly'` from `optuna.visualization`. This happened **after** the best model was saved, so nothing was lost — the model was persisted to `LightGBM_best_model.joblib` and logged to MLflow first. Only the plots and the `best_model` variable assignment were skipped. (plotly has since been installed.)
- Because the exception fired before `return best_model`, the in-memory `study` object was lost — but every trial's params and metrics are in MLflow.
- ⚠️ These params were selected by maximising accuracy **on the test set**, so `0.8629` is optimistically biased. Use a fresh held-out set for the final generalisation number.
